[EJ] 연습문제 2건

sirupe opened the pull request:
1. Multiples : 10 미만의 자연수에서 3과 5의 배수를 구하면 3, 5, 6, 9이다. 이들의 총합은 23이다. 1000 미만의 자연수에서 3, 5의 배수의 총합을 구하라.
2. Paging : 게시물의 총 건수와 한 페이지에 보여줄 게시물수를 입력으로 주었을 때 총 페이지수를 리턴하는 프로그램
로직클래스 및 테스트클래스 구현

sirupe committed:
Multiples : 10 미만의 자연수에서 3과 5의 배수를 구하면 3, 5, 6, 9이다. 이들의 총합은 23이다. 1000 미만의 자연수에서 3, 5의 배수의 총합을 구하라.
Paging : 게시물의 총 건수와 한 페이지에 보여줄 게시물수를 입력으로 주었을 때 총 페이지수를 리턴하는 프로그램 로직클래스 및 테스트클래스 구현

shinyeonjong reviewed (commented):
확인 부탁드려요

imcts commented on eunjung/test/com/java/CommonMultiplesTest.java:
```diff
@@ -0,0 +1,42 @@
+        
+        @Test
+        public void testGetResultMultipleSet() {
+            int [] resultMultiples = new int[] {3, 6, 9}; 
```
// Given 
주어지는 테스트값

// When
실행되어지는 코드

// Then
어썰쎤으로 나눌것

sirupe replied:
오.. 옙

imcts replied:
ok

sirupe replied:
수정완료

imcts commented on eunjung/test/com/java/CommonMultiplesTest.java:
```diff
@@ -0,0 +1,42 @@
+        @Test
+        public void testGetResultMultipleSet() {
+            int [] resultMultiples = new int[] {3, 6, 9}; 
+            assertArrayEquals(resultMultiples, multiples.getResultMultipleSet(startNum));
```
startNum의 경우에는 메소드 내부에서 주어지는게 명확함. 왜 저 위에서 상수처럼 선언했는지 모르겠음. 상수처럼 쓸거면 대문자로 쓰던가 상수 표기법으로

sirupe replied:
테스트코드에서도 그렇게 상수처리 해야하는지 와리가리 했음. 주어진 문제가 너무 명확해서 따로 표기하여보았음. 근데 테스트코드에 직접적으로 주어지는게 명확하겠네.

imcts replied:
상수 제거하는 편이 나음. 제거 하고 지역 변수로 수정하면 ok

sirupe replied:
수정완료

imcts commented on eunjung/test/com/java/CommonMultiplesTest.java:
```diff
@@ -0,0 +1,42 @@
+        }
+        
+        @Test
+        public void testGetResultMultipleSet() {
```
테스트메소드명은 무엇을 테스트 하고자 함인지를 명확하게 서술해야함. 이건 그냥 getResultMultipleSet을 실행하겠다는것밖에 안되잖아.

imcts replied:
자바의 테스트메소드명 참조.

sirupe replied:
오키.. 근데 IDE 에서 테스트메소드를 자동으로 만들기 하면 다 그런식이더라고 .. 아무래도 수작업이 좀 필요할 듯..?

imcts replied:
자바의 Junit 메소드 명명 규칙이 그렇다면 상관 없다고 생각함. 그래서 테스트메소드명을 참조하라고 한 것
하지만 만약 그렇다면 @Test 어노테이션에 어떤 테스트를 수행하는지는 명시하는편이 좋다고 봄. 
근데 개인적으로도 메소드명에 어떤 테스트를 수행하는지 서술하는 코드를 더 많이 봄. 

왜냐면 메소드에는 경우의 수가 하나만 있지 않을 뿐더러 (if문이 존재하는경우) 

하나의 메소드에서 그 경우를 다 테스트 하는 경우도 있지만 그 경우가 너무 많거나 복잡한 경우 
테스트 메소드도 여러개로 분기처리할 수 있다고 생각함. 어떻게 정하고 갈지는 본인 판단.

sirupe replied:
아직은 JUnit 의 사용법 정도만 알고있기 때문에 상세한 명명규칙이나 기법등은 차차 알아보도록 하겠음. 당장 찾아보았으나 명확히 이러하다.. 는 공통문서를 아직 찾지 못함.

imcts commented on eunjung/src/com/java/Multiples.java:
```diff
@@ -0,0 +1,55 @@
+    private int getMultipleSetSize(int startNumber) {
+        int setSize = 0;
+        if (startNumber > 0) {
+            setSize = maxNumber / startNumber;
```
maxNumber에 this. 추가

sirupe replied:
this 를 꼭꼭꼭꼭꼭꼭꼭 써야 할까? 시야를 많이 어지럽히기도 하더라고. 요샌 고민중인데 반드시반드시반드시반드시 필요하다고 생각한다면 수정하도록 하겠음.

imcts replied:
일단 기본적으로 문법에 명시된 것은 써야 하는게 맞는 거라고 생각함. 
시야를 어지럽히기 때문에 사용하지 않고 고민중이다? 그게 회사 모든 개발자들이 그렇다면 ok. 
하지만 this를 명시하면 시야가 어지럽고 this를 명시하지 않으면 시야가 어지럽지 않나 ? 기준은 ?
분명한 이점은 메소드가 수십줄로 길어졌을때 이 변수가 지역변수인지 아니면 자유변수인지를 판단할 수 있는
지표가 되는 키워드이기 때문에 넣는게 바람직함. 

왜 변수명을 읽을때마다 메소드 바디에서 저 변수명을 일일이 찾아야 하지 ? 

그게 정말 명시적이라고 판단하는 것인지.

sirupe replied:
수정완료

imcts commented on eunjung/src/com/java/Multiples.java:
```diff
@@ -0,0 +1,55 @@
+public class Multiples {
+    private int maxNumber;
+    
+    public Multiples(int maxNumber) {
```
인자를 받는 생성자가 있으면 인자가 없는 default생성자를 만들어주는게 좋음. 
default constructor내부에서는 this.maxNumber의 기본값을 설정하는게 제일 좋음. 
물론 여기에서는 반드시 maxNumber가 있어야 하기 때문에 만들어두었으므로 없는걸로 인지함.

sirupe replied:
이건 '물론 여기에서는 반드시 maxNumber가 있어야 하기 때문에' 라고 의도한 것이 맞음.

imcts replied:
ok

imcts commented on eunjung/src/com/java/Multiples.java:
```diff
@@ -0,0 +1,55 @@
+        this.maxNumber = maxNumber;
+    }
+    
+    public int getMaxNumber() {
```
getter는 맨 아래로

sirupe replied:
오키

imcts replied:
코드 다시 보니 getter 사용하지도 않는데 생성되어져있음. 제거.

sirupe replied:
test 코드에서 사용중. MultiplesTest 클래스 18 line.

imcts commented on eunjung/src/com/java/Multiples.java:
```diff
@@ -0,0 +1,55 @@
+        return maxNumber;
+    }
+    
+    public void setMaxNumber(int maxNumber) {
```
마찬가지

sirupe replied:
오키

imcts replied:
코드 다시 보니 setter 사용하지도 않는데 생성되어져있음. 제거.

sirupe replied:
test 코드에서 사용중. NumtiplesTest 클래스 88 line.

imcts commented on eunjung/src/com/java/Multiples.java:
```diff
@@ -0,0 +1,55 @@
+        return resultMultipleSet;
+    }
+    
+    private int getMultipleSetSize(int startNumber) {
```
대체 이 메소드가 뭐하는 메소드인지 모르겠음. 무슨 사이즈를 구한다는건지 setSize가 뭔지 모르겠음.

sirupe replied:
내가 set 을 명사로써 사용하려고 했던 의도 자체가 잘못된 듯. 네이밍을 바꿔보겠음.

imcts replied:
ok.

sirupe replied:
수정완료

imcts commented on eunjung/src/com/java/Multiples.java:
```diff
@@ -0,0 +1,55 @@
+        int setSize = 0;
+        if (startNumber > 0) {
+            setSize = maxNumber / startNumber;
+            if (maxNumber % setSize == 0) {
```
여기서는 대체 왜 --;하고?

imcts commented on eunjung/src/com/java/Multiples.java:
```diff
@@ -0,0 +1,55 @@
+        
+        int multipleNumber = startNumber;
+        for (int i = 0; i < setSize; i++) {
+            resultMultipleSet[i] = (multipleNumber);
```
`(multipleNumber);` 이게 대체..

sirupe replied:
That's my mistake....

imcts replied:
ok.

imcts commented on eunjung/src/com/java/Multiples.java:
```diff
@@ -0,0 +1,55 @@
+    }
+    
+    public int [] getResultMultipleSet(int startNumber) {
+        int setSize = getMultipleSetSize(startNumber);
```
set은 설정하다라는 동사일수도 있고 자바문법에서 사용되는 컬렉션 이름이기도 함. 
그럼 대체 무슨 셋의 사이즈를 말한다는건지도 모르겠고 실질적으로 만들어지는건 심지어 배열이야. 결국 for문의 limit값을 구한다는 거면 이게 이렇게 네이밍 되어져야 할 필요가 있을까.

sirupe replied:
그냥 size 여도 괜찮을까?

imcts replied:
코드 보니까 그냥 만들고 싶은 배열의 사이즈를 얻고 싶은것 같은데 그냥 `getMultipleArrayLength`여도 충분할듯. size가 아니잖아 배열이니까. 

근데 근본적으로 어제 말했듯이 굳이 배열을 써야하나? 굳이 쓸 필요 없는데 쓰고있는 기분이 듬.

sirupe replied:
1. 배열의 사이즈를 얻는 코드를 제거함.
2. 배열 -> Set 으로 변경함

imcts commented on eunjung/src/com/java/Multiples.java:
```diff
@@ -0,0 +1,55 @@
+    
+    public int [] getResultMultipleSet(int startNumber) {
+        int setSize = getMultipleSetSize(startNumber);
+        int [] resultMultipleSet = new int[setSize];
```
굳이 배열을 써야 하는 이유는 또 뭐야 어차피 setSize값만큼만 포문 돌면서 집어 넣을거면 list가 더 나은거 아닌가

sirupe replied:
배열 -> Set 으로 변경함

imcts commented on eunjung/src/com/java/Multiples.java:
```diff
@@ -0,0 +1,55 @@
+                setSize--;
+            }
+        }
+        return setSize;
```
이 메소드 자체가 너무 의미 모호함. 좀더 명확하게 기술해줬으면 좋을듯

sirupe replied:
배열을 이용했었기 때문에 사이즈를 구하는 로직이 필요했었는데 자료형을 Set 으로 변경함으로 인해 해당 메소드를 삭제하였음.

imcts commented on eunjung/src/com/java/Multiples.java:
```diff
@@ -0,0 +1,55 @@
+        int [] firstNumberMultipleSet = getResultMultipleSet(firstNumber);
+        int [] secondNumberMultipleSet = getResultMultipleSet(secondNumber);
+    
+        int firstSumNumberOfMultiple = Arrays.stream(firstNumberMultipleSet).sum();
```
3, 5의 배수의 총합을 구하라면서 여기는 왜 더하지? 필터링 하지 않고

sirupe replied:
3, 5의 배수의 총합을 구한다는 건 '3의 배수의 합' + '5의 배수의 합' 에서 중복되어 더해지는 값만 제외하면 되는 것으로 두 수 중 하나는 온전히 더하면 되고 그 공배수의 수를 제외해주는 작업은 다른 한 배수의 합을 구할때만 이루어지면 됨. 그래서 하나의 수는 온전히 더했고 하나의 수는 공배수를 한 번 제외해주는 작업을 추가하였음. 흐름상 이상하다면 답변 바람.

imcts replied:
그건 본인만 보고 판단할 수 있는 코드일뿐더러 어제 내가 물어봤을때 `어? 이건 왜 이렇게 했지?? 기억이 안나는데?` 라는 답변을 한 것 자체가 이상한 것임. 본인이 작성한 코드를 읽지 못하는데 남이 그 코드를 어떻게 읽고 판단하라는 이야기임. 만약 메소드 자체가 너무 복잡하여 도저히 이해할 수 없다면 메소드 위에 이 메소드가 하는 일을 서술하는 주석이라도 달아야 함. 자바 주석 다는 방법은 구글 검색 참조. 

그리고 추가로 하나 더 말하자면 `getResultMultipleSet`이게 맞아 ?
내 보기엔 `getMultipleValues`가 맞는거 같은데. 추가해서 계산된정도 붙여준다면 `getCalculatedValues`정도면 알아보기 쉬울듯.

sirupe replied:
1. 중복되는 공배수가 두 번 들어가지 않도록 하는 작업에서 제외하는 로직을 추가하는 것이 아닌 Set 자료형을 사용하는 것으로 변경하였음.
2. 메소드명은 getResultMultipleValues 로 변경하였음. (인자값의 배수집단을 구하는 메소드이기에 getCalculatedValues 는 너무 포괄적인 의미라고 생각함.)

imcts commented on eunjung/src/com/java/Paging.java:
```diff
@@ -0,0 +1,16 @@
+//  입력 : 총건수(m), 한페이지에 보여줄 게시물수(n) (단 n은 1보다 크거나 같다. n >= 1)
+//  출력 : 총페이지수
+    public int getTotalPageCount(int totalListCount, int countOfOnePage) {
+        if (!(countOfOnePage >= 1)) {
```
```java
 if (countOfOnePage <= 1) {
     return 0; 
 }
```

???????

sirupe replied:
웜.. 이건 너무 의식의 흐름이어찌? 수정하도록 할게

imcts replied:
ok

imcts reviewed (changes requested):
배열은 리스트로 변경 하면 더 좋을듯. 
그리고 메소드명이 명확하지 않고 무슨 일을 하는지 명확하게 보여주지 않음. Multiple은 리팩토링 요망.

Commit:
Merge branch 'develop' into feature/eunjung

Commit:
- git ignore 추가.

Commit:
- git ignore 추가.

Commit:
- git ignore 추가.

Commit:
- git ignore 추가.

imcts commented on eunjung/eunjung.iml:
```diff
@@ -1,8 +1,11 @@
 <?xml version="1.0" encoding="UTF-8"?>
-<module type="WEB_MODULE" version="4">
+<module type="JAVA_MODULE" version="4">
```
테스트용 리뷰 1

imcts commented on eunjung/eunjung.iml:
```diff
@@ -1,8 +1,11 @@
   <component name="NewModuleRootManager" inherit-compiler-output="true">
     <exclude-output />
-    <content url="file://$MODULE_DIR$" />
+    <content url="file://$MODULE_DIR$">
```
테스트용 리뷰 2

imcts reviewed (commented):
리뷰 확인 부탁 드려요.

sirupe committed:
Merge branch 'develop' of github.com:howCanWeDoVeryWell/CodeReview into feature/eunjung

sirupe committed:
1. CommonMultiplesTest.java -> MultipleTest.java : 클래스 이름 변경
2. MultiplesTest.java : 상수를 제거하고 지역변수로 변경.
3. Multiples.java
   * 인스턴스 변수에 this. 붙임.
   * getResultMultipleSet() -> getResultMultipleValues() 로 이름 변경
   * getResultMultipleValues() 내부의 배수 집합 자료형을 `int []` 에서 `Set` 으로 변경

sirupe committed:
1. CommonMultiplesTest.java -> MultipleTest.java : 클래스 이름 변경
2. MultiplesTest.java : 상수를 제거하고 지역변수로 변경.
3. Multiples.java
   * 인스턴스 변수에 this. 붙임.
   * getResultMultipleSet() -> getResultMultipleValues() 로 이름 변경
   * getResultMultipleValues() 내부의 배수 집합 자료형을 `int []` 에서 `Set` 으로 변경

sirupe committed:
imi added.

sirupe committed:
imi delete

sirupe committed:
Merge branch 'feature/eunjung' of github.com:howCanWeDoVeryWell/CodeReview into feature/eunjung

sirupe committed:
imi delete

sirupe committed:
imi delete

imcts commented on eunjung/src/com/java/Multiples.java:
```diff
@@ -0,0 +1,42 @@
+        return resultMultipleSet;
+    }
+    
+    public int getMaxNumber() {
```
`getResultMultipleValues`에 달아놓은 코멘트를 기준으로 설명하면 이 getter, setter또한 있어서는 안되는 메소드임.

sirupe replied:
getter, setter 삭제 및 getter, setter 로 test 에서 테스트 중이었던 내용 삭제.

imcts replied:
ok

imcts commented on eunjung/src/com/java/Multiples.java:
```diff
@@ -0,0 +1,42 @@
+        return multipleSet.stream().reduce((a, b) -> a + b).get();
+    }
+    
+    public Set<Integer> getResultMultipleValues(int... numbers) {
```
`getResultMultipleValues`는 public이어야할 이유가 없음. 메소드를 외부로 노출 할 것이냐 말 것이냐를 결정하는 건
이 클래스의 API를 결정하는 사안이기때문에 API가 많이 외부로 노출되는 클래스는 좋은 클래스가 아님. 
즉 외부에서 접근할 수 있는 메소드가 많다라는 것 자체가 좋은게 아닐 뿐더러 `getResultMultipleValues`는 단순히 
`getSumOfMultipleOfTwoNumbers `에서 분리되어진 메소드로 보여지기 떄문에 private 여야 함. 

`어? 그럼 테스트는 어떻게 하라는거야`라는 의문이 생긴다면 테스트의 범위를 잘 모르고 있기 때문임
즉 Multiples에서 외부로 노출해야하는 메소드는 오로지 getSumOfMultipleOfTwoNumbers 하나 밖에 없고
모든 테스트는 이 메소드를 통해서 수행해야 하는 것.

sirupe replied:
접근제한자 및 관련 테스트케이스 삭제.

imcts replied:
ok

imcts commented on eunjung/src/com/java/Multiples.java:
```diff
@@ -0,0 +1,42 @@
+        this.maxNumber = maxNumber;
+    }
+    
+    public int getSumOfMultipleOfTwoNumbers(int firstNumber, int secondNumber) {
```
두개의 숫자들의 곱셈의 합이라는 소유격을 써야할 이유가 뭐가 있는지를 모르겠음. 
전치사를 메소드명에 바르기 시작하면 끝도 없기 때문에 최대한 지양하는게 좋을뿐더러 쓴다 하더라도 요새는 
한두개정도만 사용. 오히려 가독성이 더 떨어지는 결과를 초래함.

`getMultipleSumBy`정도면 되는거시 아닐까. 이름을 좀 간단하고 명시적이게 다시 짓는걸 추천함.

sirupe replied:
수정함.

imcts replied:
ok

imcts commented on eunjung/src/com/java/Multiples.java:
```diff
@@ -0,0 +1,42 @@
+        Set<Integer> resultMultipleSet = new HashSet<>();
+        Arrays.stream(numbers).forEach(number -> {
+            if (number != 0) {
+                for (int i = number; i < maxNumber; i += number) {
```
`i < maxNumber`가 아니라 `i <= maxNumber`여야 맞지 않을까. 
그래야 1000과 같거나 작은 경우를 볼거 같은데. 1000은 5의 배수가 아닌가?

sirupe replied:
문제가 1000 미만이기 때문.

imcts replied:
ok

imcts commented on eunjung/src/com/java/Paging.java:
```diff
@@ -0,0 +1,16 @@
+//  입력 : 총건수(m), 한페이지에 보여줄 게시물수(n) (단 n은 1보다 크거나 같다. n >= 1)
+//  출력 : 총페이지수
+    public int getTotalPageCount(int totalListCount, int countOfOnePage) {
+        if (!(countOfOnePage >= 1)) {
```
```java
 if (countOfOnePage <= 1) {
     return 0; 
 }
```
여전히 수정되지 않았음.

sirupe replied:
수정함

imcts replied:
ok

imcts reviewed (commented):
확인해 주셈.

sirupe committed:
Multiples
1. 메소드명 수정 (getSumOfMultipleOfTwoNumbers -> getMultipleSumBy)
2. getResultMultipleValues 메소드 접근제한자 수정 (public -> private)
3. maxNumger getter setter 삭제
4. Multiple 변경사항에 따른 테스트케이스 삭제
Paging
1. getTotalPageCount 조건문 조건절 변경

imcts reviewed (approved):
ok
